pywbem_mock/_defaultinstanceprovider.py

```diff
@@ -0,0 +1,527 @@
+        if classname_lower == 'pg_namespace':
+            ns_classname = 'PG_Namespace'
+        elif classname_lower == 'cim_namespace':
+            ns_classname = 'CIM_Namespace'
+        else:
+            ns_classname = None
+        if ns_classname:
+            try:
+                new_namespace = new_instance['Name']
+            except KeyError:
+                raise CIMError(
+                    CIM_ERR_INVALID_PARAMETER,
+                    _format("Namespace creation via CreateInstance: "
+                            "Missing 'Name' property in the {0!A} instance ",
+                            new_instance.classname))
+
+            # Normalize the namespace name
+            # TODO: We should not have to normalize here, that is lower level
+            new_namespace = new_namespace.strip('/')
+
+            # Write it back to the instance in case it was changed
+            new_instance['Name'] = new_namespace
+
+            # These values must match those in
+            # tests/unittest/utils/wbemserver_mock.py.
+            new_instance['CreationClassName'] = ns_classname
+            new_instance['ObjectManagerName'] = 'MyFakeObjectManager'
+            new_instance['ObjectManagerCreationClassName'] = \
+                'CIM_ObjectManager'
+            new_instance['SystemName'] = 'Mock_Test_WBEMServerTest'
+            new_instance['SystemCreationClassName'] = 'CIM_ComputerSystem'
+
+        # Test all key properties in instance. The CIM repository
+        # cannot add values for key properties and does
+        # not allow creating key properties from class defaults.
+        key_props = [p.name for p in six.itervalues(target_class.properties)
+                     if 'key' in p.qualifiers]
+        for pn in key_props:
+            if pn not in new_instance:
+                raise CIMError(
+                    CIM_ERR_INVALID_PARAMETER,
+                    _format("Key property {0!A} not in NewInstance ", pn))
+
+        # Exception if property in instance but not class or types do not
+        # match
+        for iprop_name in new_instance:
+            if iprop_name not in target_class.properties:
+                raise CIMError(
+                    CIM_ERR_INVALID_PARAMETER,
+                    _format("Property {0!A} specified in NewInstance is not "
+                            "exposed by class {1!A} in namespace {2!A}",
+                            iprop_name, target_class.classname, namespace))
+
+            cprop = target_class.properties[iprop_name]
+            iprop = new_instance.properties[iprop_name]
+            if iprop.is_array != cprop.is_array or \
+                    iprop.type != cprop.type:
+                raise CIMError(
+                    CIM_ERR_INVALID_PARAMETER,
+                    _format("Instance and class property {0!A} types "
+                            "do not match: instance={1!A}, class={2!A}",
+                            iprop_name, iprop, cprop))
```

Comment:
This check could be done before calling the provider.

Reply:
Moved to ProviderDispatcher